### PR TITLE
(TK-179) Modify the Clojure implementation to share more of the Java impl

### PR DESCRIPTION
**Pull request by scottyw**
This commit removes duplicate functionality from the Clojure implementation and reuses the Java request/response handling impl. It doesn't address client creation so there are still two distinct paths for that.

The main changes are Ring <-> Java request/response conversion and the addition of interface ResponseDeliveryDelegate. The delegate implementations take care of the language-specific aspects of response building and promise delivery. One of the Clojure test namespaces is removed too.

### CODE REVIEW

File: test/puppetlabs/http/client/async_unbuffered_test.clj [hunk 1]
```diff
@@ -180,8 +180,7 @@
   transmission"
   [decompress-body?]
   (testlogging/with-test-logging
-    (let [data (generate-data (* 32 1024 1024))
-          opts {:as :unbuffered-stream :decompress-body decompress-body?}]
+    (let [data (generate-data (* 32 1024 1024))]
```
Comment (rlinehan) [inline test/puppetlabs/http/client/async_unbuffered_test.clj]: This was just removed because it's not used in this test, right?

Comment (scottyw) [inline test/puppetlabs/http/client/async_unbuffered_test.clj]: Yes it's a rogue fix because I noticed it served no purpose.

Comment (rlinehan) [inline test/puppetlabs/http/client/async_unbuffered_test.clj]: :+1:

File: src/clj/puppetlabs/http/client/async.clj [hunk 2]
```diff
@@ -326,39 +137,80 @@
...
+  [opts :- common/UserRequestOptions]
+  (case (:method opts)
+    :delete HttpMethod/DELETE
+    :get HttpMethod/GET
+    :head HttpMethod/HEAD
+    :options HttpMethod/OPTIONS
+    :patch HttpMethod/PATCH
+    :post HttpMethod/POST
+    :put HttpMethod/PUT
+    :trace HttpMethod/TRACE))
```
Comment (rlinehan) [inline src/clj/puppetlabs/http/client/async.clj]: I noticed that previously we would throw an error if the method wasn't found (https://github.com/puppetlabs/clj-http-client/pull/47/files#diff-f2e3a844d8f57439d128ef47be66d78dL163), but with this now we don't. Should we throw an error with a better message (otherwise it would be `java.lang.IllegalArgumentException: No matching clause:`)? Is that even a case that we can reach? Does it matter that much? (there's no test I can find)

Comment (scottyw) [inline src/clj/puppetlabs/http/client/async.clj]: Good call - very easy to typo it and in light of Alex's PR to expose ```request``` directly I think we should address it. I'll hit it on Monday.

File: test/puppetlabs/http/client/async_test.clj [hunk 3]
```diff
@@ -1,37 +0,0 @@
...
-                the given mime-type but adds UTF-8 as the charset"
-        (let [body "foo"]
-          (is (= (compute-content-type body "text/html")
-                 "text/html; charset=UTF-8")))))))
-
-(deftest nil-response-body-coerced-as-text
-  (testing "a nil response body is coerced into a string by async/coerce-body-type"
-    (let [resp {:body nil, :opts {:as :text}}]
-      (is (= {:body "", :opts {:as :text}}
-             (coerce-body-type resp))))))
```
Comment (rlinehan) [inline test/puppetlabs/http/client/async_test.clj]: I feel like removing these tests (and the accompanying code that does the coercion of content type) might be a little dangerous (without adding something comparable in java to achieve this result)? I think that this coercion was specifically added - e.g. TK-145/SERVER-294 and TK-182.

Comment (scottyw) [inline test/puppetlabs/http/client/async_test.clj]: My reading of the code is that these tests are the Clojure version of these Java tests:

https://github.com/puppetlabs/clj-http-client/blob/master/test/com/puppetlabs/http/client/impl/java_client_test.clj

Those two Clojure functions under test have been removed too, with the functionality that was in ```content-type``` now here:

https://github.com/puppetlabs/clj-http-client/blob/master/src/java/com/puppetlabs/http/client/impl/JavaClient.java#L91

Ditto ```coerce-body-type```:

https://github.com/puppetlabs/clj-http-client/blob/master/src/java/com/puppetlabs/http/client/impl/JavaClient.java#L473

There are some end-to-end tests here that I think might hit this stuff?

https://github.com/puppetlabs/clj-http-client/blob/master/test/puppetlabs/http/client/sync_plaintext_test.clj#L295

What do you think?

Comment (rlinehan) [inline test/puppetlabs/http/client/async_test.clj]: I swear I searched the java code for the equivalent `content-type` munging as was in the clojure code. Obviously I shouldn't review PRs while sick. Yeah, totally agree with you that all of the stuff you linked to is sufficient, and I'm :+1: on this.

**Comment by rlinehan** [comment]
I'm :+1: on this aside from adding handling for an incorrect http method. Would be good for someone more familiar with java than me to review (@cprice404).

File: src/clj/puppetlabs/http/client/async.clj (hunk 2 above)
Comment (scottyw) [inline src/clj/puppetlabs/http/client/async.clj]: Addressed in a second commit.

I didn't add an explicit test - there seemed no good place bar a dedicated ```clojure-method->java``` unit test which seemed a bit redundant.

Comment (rlinehan) [inline src/clj/puppetlabs/http/client/async.clj]: Okay, that seems fine. Seems like it might be more possible to add a test after Alex's request PR is merged, but doesn't seem entirely necessary or worth the effort.

File: src/java/com/puppetlabs/http/client/impl/JavaClient.java [hunk 4]
```diff
@@ -307,35 +309,34 @@ public void cancelled() {
...
             @Override
             public void completed(HttpResponse httpResponse) {
-                completeResponse(promise, requestOptions, callback, httpResponse, httpContext);
+                completeResponse(responseDeliveryDelegate, requestOptions, callback, httpResponse, httpContext);
             }
 
             @Override
             public void failed(Exception e) {
-                deliverResponse(requestOptions, new Response(requestOptions, e), callback, promise);
+                responseDeliveryDelegate.deliverResponse(requestOptions, e);
```
Comment (camlow325) [inline src/java/com/puppetlabs/http/client/impl/JavaClient.java]: If a callback was provided for the request, will it still be invoked if the request fails?

File: src/java/com/puppetlabs/http/client/impl/JavaClient.java [hunk 5]
```diff
@@ -307,35 +309,34 @@ public void cancelled() {
...
             }
 
             @Override
             public void cancelled() {
-                deliverResponse(requestOptions, new Response(requestOptions, new HttpClientException("Request cancelled", null)), callback, promise);
+                responseDeliveryDelegate.deliverResponse(requestOptions, new HttpClientException("Request cancelled"));
```
Comment (camlow325) [inline src/java/com/puppetlabs/http/client/impl/JavaClient.java]: Same question here on the callback as with the `failed()` method above.

File: src/clj/puppetlabs/http/client/async.clj [hunk 6]
```diff
@@ -12,27 +12,17 @@
...
+           (com.puppetlabs.http.client.impl JavaClient ResponseDeliveryDelegate)
            (org.apache.http.client RedirectStrategy)
            (org.apache.http.impl.client LaxRedirectStrategy DefaultRedirectStrategy)
            (org.apache.http.nio.conn.ssl SSLIOSessionStrategy)
            (org.apache.http.client.config RequestConfig)
-           (org.apache.http.nio.client.methods HttpAsyncMethods)
            (org.apache.http.nio.client HttpAsyncClient)
-           (org.apache.http.client.protocol ResponseContentEncoding HttpClientContext)
-           (org.apache.http.protocol HttpContext))
+           (org.apache.http.entity ContentType))
```
Comment (camlow325) [inline src/clj/puppetlabs/http/client/async.clj]: Is this import used?

File: src/clj/puppetlabs/http/client/async.clj [hunk 7]
```diff
@@ -326,39 +137,81 @@
...
+  (if callback
+    (try
+      (callback response)
+      (catch Exception e
+        {:opts opts :error e}))
+    response))
+
+(schema/defn java-content-type->clj :- common/ContentType
+  [java-content-type]
+  (if java-content-type
```
Comment (camlow325) [inline src/clj/puppetlabs/http/client/async.clj]: It appears that this implementation would cause a response with an HTTP header of "Content-Type=" (note no value to the right of the "=" sign) to throw an exception back to the caller.  I got this when I tried:

~~~~
ParseException Invalid content type:   org.apache.http.entity.ContentType.parse (ContentType.java:273)
~~~~

The old implementation does an `if (empty? ...` check on the content-type before trying to use it and, as a result, delivers a Content-Type of `nil` with an otherwise full response body.

Arguably, I suppose, this kind of header value might be considered malformed.  Since it's possible for a server to do this, though, I think it would be good to preserve the spirit of the error handling that the old implementation did for this case.

Comment (camlow325) [inline src/clj/puppetlabs/http/client/async.clj]: Should probably add a test that exercises this too since none of the others appear to have caught it?

**Comment by camlow325** [comment]
Overall :+1: Looking really good to me.  Very happy to see this work finally happening.

I just left a few comments throughout.  The main concerns I have are around whether or not the callback is being exercised in all cases now and with the one minor issue with an empty "Content-Type" header in a response.

**Comment by scottyw** [comment]
@camlow325 Thanks vey much!

**Comment by scottyw** [comment]
@rlinehan Added a test to check behaviour with a bad HTTP verb

File: src/clj/puppetlabs/http/client/async.clj (hunk 7 above)
Comment (scottyw) [inline src/clj/puppetlabs/http/client/async.clj]: Added the test and updated the Java content type parsing to check for empty. In the last commit.

**Comment by scottyw** [comment]
Updating the callback behaviour. Thanks for the spot. Again in a separate commit.

**Comment by scottyw** [comment]
@camlow325 @rlinehan Thanks for the help

**Comment by rlinehan** [comment]
I think it would be good to have a test of the callback behavior, but as there weren't any previously I'm cool with that being addressed in a separate PR.

:+1: :shipit:

**Comment by scottyw** [comment]
Fantastic - thanks @rlinehan and @camlow325!
